### PR TITLE
Added feature to download expansion files if deleted by user

### PR DESCRIPTION
**Pull request by akashdabhi03**
|PR checklist|Status|
|-----|:------:|
|Is code analysis task executed and issues resolved in the Java/Kotlin code?|Yes|
|Is code analysis task executed and issues resolved in the CPP/native code?|NA|
|Is the version code set in the AndroidManifest.xml or build.gradle file to 1 more than the version which has already been published?|Yes|
|Is code commenting done properly for the changes made?|Yes|
|Is correct variable naming convention followed?|Yes|
|Is auto formatting shortcut Ctrl+Alt+Shift+L executed in all files?|Yes|
|Are the message strings used in the application taken from strings.xml file?|Yes|
|Are the permissions which are not required removed from AndroidManifest.java file?|NA|
|Is the _uses-feature_ tag used for only must have feature in AndroidManifest.java file?|NA|
|Are the generated files correctly configured in .gitignore?|NA|
|Are the release signing configurations removed from the gradle files?|NA|
|Is commented code removed from all files?|Yes|
|Are updated images compressed (< 100 KB)?|NA|
|Ensured there are no typos in any string. For example: Variable/method names, hardcoded strings, commit messages, PR titles, issues, code comments?|Yes|

OBB Checklist:  

|PR checklist|Status|
|-----|:------:|
|Is size of the OBB correct?|Yes|
|Is OBB file version correctly set in DownloadExpansionFile.java file?|Yes|
|Is patch OBB file to be deployed? If yes, then are the settings correct in DownloadExpansionFile.java file?|No|

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -33,8 +33,8 @@ android {
         applicationId 'com.maq.xprize.onecourse.hindi'
         minSdkVersion 23
         targetSdkVersion 28
-        versionCode 12
-        versionName "4.0.6"
+        versionCode 14
+        versionName "4.0.8"
```
Comment (ankitk20) [inline app/build.gradle]: Update version name as per convention

File: app/src/main/java/com/maq/xprize/onecourse/hindi/utils/Zip.java [hunk 2]
```diff
@@ -100,19 +101,23 @@ public void run() {
...
                     inputStream.close();
                 }
             }
         }
+        if (isExpansionSuccessful) {
+            if (isMain) {
+                editor.putInt("mainFileVersion", expansionFileVersion);
+            } else {
+                editor.putInt("patchFileVersion", expansionFileVersion);
+            }
```
Comment (ankitk20) [inline app/src/main/java/com/maq/xprize/onecourse/hindi/utils/Zip.java]: __editor.putInt(isMain? "mainFileVersion" : "patchFileVersion", expansionFileVersion)__
User ternary operator for readability

File: downloader_library/src/main/java/com/google/android/vending/expansion/downloader/Helpers.java [hunk 3]
```diff
@@ -209,54 +215,97 @@ public static String getExpansionAPKFileName(Context c, boolean mainFile, int ve
...
+     * Returns the filename (where the file should be saved) from info about a download
      */
     static public String generateSaveFileName(Context c, String fileName) {
-        String path = getSaveFilePath(c)
+        return getSaveFilePath(c)
                 + File.separator + fileName;
-        return path;
     }
 
+    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
```
Comment (ankitk20) [inline downloader_library/src/main/java/com/google/android/vending/expansion/downloader/Helpers.java]: Check if this is TargetApi is required anymore

File: downloader_library/src/main/java/com/google/android/vending/expansion/downloader/impl/DownloadNotification.java [hunk 4]
```diff
@@ -41,37 +45,79 @@
...
     private DownloadProgressInfo mProgressInfo;
+    private PendingIntent mContentIntent;
+
+    /**
+     * Constructor
+     *
+     * @param ctx The context to use to obtain access to the Notification
+     *            Service
+     */
+    @RequiresApi(api = Build.VERSION_CODES.O)
```
Comment (ankitk20) [inline downloader_library/src/main/java/com/google/android/vending/expansion/downloader/impl/DownloadNotification.java]: Check if this is required

File: downloader_library/src/main/java/com/google/android/vending/expansion/downloader/impl/DownloadNotification.java [hunk 5]
```diff
@@ -130,22 +176,28 @@ public void onDownloadStateChanged(int newState) {
...
             } else {
-                mCurrentNotification.flags &= ~Notification.FLAG_ONGOING_EVENT;
-                mCurrentNotification.flags |= Notification.FLAG_AUTO_CANCEL;
+                mCurrentBuilder.setOngoing(false);
+                mCurrentBuilder.setAutoCancel(true);
+            }
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+                notificationManager.notify(NOTIFICATION_ID, mCurrentBuilder.build());
+            } else {
+                mNotificationManager.notify(NOTIFICATION_ID, mCurrentBuilder.build());
```
Comment (ankitk20) [inline downloader_library/src/main/java/com/google/android/vending/expansion/downloader/impl/DownloadNotification.java:195]: Check this _if-else_

File: downloader_library/src/main/java/com/google/android/vending/expansion/downloader/impl/DownloadNotification.java [hunk 6]
```diff
@@ -130,22 +176,28 @@ public void onDownloadStateChanged(int newState) {
...
             }
-            mNotificationManager.notify(NOTIFICATION_ID, mCurrentNotification);
         }
     }
 
+    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN)
```
Comment (ankitk20) [inline downloader_library/src/main/java/com/google/android/vending/expansion/downloader/impl/DownloadNotification.java]: Check

File: downloader_library/src/main/java/com/google/android/vending/expansion/downloader/impl/DownloadNotification.java [hunk 7]
```diff
@@ -154,48 +206,35 @@ public void onDownloadProgress(DownloadProgressInfo progress) {
...
+            mActiveDownloadBuilder.setSmallIcon(android.R.drawable.stat_sys_download);
+            mActiveDownloadBuilder.setTicker(mLabel + ": " + mCurrentText);
+            mActiveDownloadBuilder.setContentTitle(mLabel);
+            mActiveDownloadBuilder.setContentInfo(mContext.getString(R.string.time_remaining_notification,
+                    Helpers.getTimeRemaining(progress.mTimeRemaining)));
+            mCurrentBuilder = mActiveDownloadBuilder;
+        }
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            notificationManager.notify(NOTIFICATION_ID, mCurrentBuilder.build());
+        } else {
```
Comment (ankitk20) [inline downloader_library/src/main/java/com/google/android/vending/expansion/downloader/impl/DownloadNotification.java:225]: Check

**Review by ankitk20 — CHANGES_REQUESTED** [review]

File: app/build.gradle (hunk 1 above)
Comment (akashdabhi03) [reply app/build.gradle]: Updated

File: app/src/main/java/com/maq/xprize/onecourse/hindi/utils/Zip.java (hunk 2 above)
Comment (akashdabhi03) [reply app/src/main/java/com/maq/xprize/onecourse/hindi/utils/Zip.java]: Done

File: app/build.gradle (hunk 1 above)
Comment (akashdabhi03) [reply app/build.gradle]: Updated

File: downloader_library/src/main/java/com/google/android/vending/expansion/downloader/Helpers.java (hunk 3 above)
Comment (akashdabhi03) [reply downloader_library/src/main/java/com/google/android/vending/expansion/downloader/Helpers.java]: Removed

File: downloader_library/src/main/java/com/google/android/vending/expansion/downloader/impl/DownloadNotification.java (hunk 4 above)
Comment (akashdabhi03) [reply downloader_library/src/main/java/com/google/android/vending/expansion/downloader/impl/DownloadNotification.java]: Removed

File: downloader_library/src/main/java/com/google/android/vending/expansion/downloader/impl/DownloadNotification.java (hunk 5 above)
Comment (akashdabhi03) [reply downloader_library/src/main/java/com/google/android/vending/expansion/downloader/impl/DownloadNotification.java:195]: Yes this is required to make the notification compatible across Android versions

File: downloader_library/src/main/java/com/google/android/vending/expansion/downloader/impl/DownloadNotification.java (hunk 6 above)
Comment (akashdabhi03) [reply downloader_library/src/main/java/com/google/android/vending/expansion/downloader/impl/DownloadNotification.java]: Removed

File: downloader_library/src/main/java/com/google/android/vending/expansion/downloader/impl/DownloadNotification.java (hunk 7 above)
Comment (akashdabhi03) [reply downloader_library/src/main/java/com/google/android/vending/expansion/downloader/impl/DownloadNotification.java:225]: Required
Same reason as above

**Review by ankitk20 — DISMISSED** [review]

**Review by ankitk20 — DISMISSED** [review]

**Review by Shailin — DISMISSED** [review]

**Review by Shailin — APPROVED** [review]

**Review by ankitk20 — APPROVED** [review]

**Review by mathewnik90 — APPROVED** [review]

**Pull request merged**